File: src/api/social.ts
                                                                                                                                                                                                                                                                                                                       import request from '@/utils/request'

// 动态广场相关接口
export interface Post {
  id: number
  user_id: number
  user?: {
    id: number
    username: string
    nickname: string
    avatar?: string
  }
  post_type: 'text' | 'image' | 'video' | 'dance' | 'heritage'
  content: string
  media_url?: string
  likes_count: number
  comments_count: number
  shares_count: number
  is_public: boolean
  is_featured: boolean
  created_at: string
  updated_at: string
}

export interface PostCreate {
  post_type: 'text' | 'image' | 'video' | 'dance' | 'heritage'
  content: string
  media_url?: string
  is_public?: boolean
}

export interface PostComment {
  id: number
  post_id: number
  user_id: number
  user?: {
    id: number
    username: string
    nickname: string
    avatar?: string
  }
  content: string
  parent_id?: number
  created_at: string
}

export interface CommentCreate {
  content: string
  parent_id?: number
}

// 非遗传承相关接口
export interface HeritageProject {
  id: number
  name: string
  category: string
  level: 'national' | 'provincial' | 'municipal' | 'county'
  description: string
  history: string
  techniques: string
  current_status: string
  protection_measures: string
  image_url?: string
  video_url?: string
  status: 'active' | 'inactive'
  created_at: string
  updated_at: string
}

export interface HeritageInheritor {
  id: number
  name: string
  gender: 'male' | 'female'
  birth_year: number
  hometown: string
  bio: string
  achievements: string
  inheritance_years: number
  teaching_experience: string
  representative_works: string
  contact_info: string
  avatar_url?: string
  status: 'active' | 'inactive'
  created_at: string
  updated_at: string
}

// 动态广场API
export const postApi = {
  // 获取动态列表
  getPosts: (params: {
    page?: number
    page_size?: number
    post_type?: string
    user_role?: string
    is_public?: boolean
    is_featured?: boolean
  }) => {
    return request.get('/api/v1/social/posts', { params })
  },

  // 创建动态
  createPost: (data: PostCreate) => {
    return request.post('/api/v1/social/posts', data)
  },

  // 获取动态详情
  getPost: (id: number) => {
    return request.get(`/api/v1/social/posts/${id}`)
  },

  // 点赞动态
  likePost: (id: number) => {
    return request.post(`/api/v1/social/posts/${id}/like`)
  },

  // 取消点赞
  unlikePost: (id: number) => {
    return request.delete(`/api/v1/social/posts/${id}/like`)
  },

  // 获取动态评论
  getComments: (postId: number, params?: {
    page?: number
    page_size?: number
  }) => {
    return request.get(`/api/v1/social/posts/${postId}/comments`, { params })
  },

  // 添加评论
  addComment: (postId: number, data: CommentCreate) => {
    return request.post(`/api/v1/social/posts/${postId}/comments`, data)
  },

  // 删除评论
  deleteComment: (postId: number, commentId: number) => {
    return request.delete(`/api/v1/social/posts/${postId}/comments/${commentId}`)
  },

  // 切换动态精选状态（管理员）
  toggleFeatured: (id: number) => {
    return request.patch(`/api/v1/social/posts/${id}/featured`)
  },

  // 切换动态可见性（管理员）
  toggleVisibility: (id: number) => {
    return request.patch(`/api/v1/social/posts/${id}/visibility`)
  },

  // 删除动态（管理员）
  deletePost: (id: number) => {
    return request.delete(`/api/v1/social/posts/${id}`)
  }
}

// 非遗项目API
export const heritageProjectApi = {
  // 获取项目列表
  getProjects: (params: {
    page?: number
    page_size?: number
    keyword?: string
    category?: string
    level?: string
    status?: string
  }) => {
    return request.get('/api/v1/social/heritage/projects', { params })
  },

  // 创建项目
  createProject: (data: Partial<HeritageProject>) => {
    return request.post('/api/v1/social/heritage/projects', data)
  },

  // 获取项目详情
  getProject: (id: number) => {
    return request.get(`/api/v1/social/heritage/projects/${id}`)
  },

  // 更新项目
  updateProject: (id: number, data: Partial<HeritageProject>) => {
    return request.put(`/api/v1/social/heritage/projects/${id}`, data)
  },

  // 切换项目状态
  toggleStatus: (id: number) => {
    return request.patch(`/api/v1/social/heritage/projects/${id}/status`)
  },

  // 删除项目
  deleteProject: (id: number) => {
    return request.delete(`/api/v1/social/heritage/projects/${id}`)
  }
}

// 非遗传承人API
export const heritageInheritorApi = {
  // 获取传承人列表
  getInheritors: (params: {
    page?: number
    page_size?: number
    keyword?: string
    hometown?: string
    gender?: string
    status?: string
  }) => {
    return request.get('/api/v1/social/heritage/inheritors', { params })
  },

  // 创建传承人
  createInheritor: (data: Partial<HeritageInheritor>) => {
    return request.post('/api/v1/social/heritage/inheritors', data)
  },

  // 获取传承人详情
  getInheritor: (id: number) => {
    return request.get(`/api/v1/social/heritage/inheritors/${id}`)
  },

  // 更新传承人
  updateInheritor: (id: number, data: Partial<HeritageInheritor>) => {
    return request.put(`/api/v1/social/heritage/inheritors/${id}`, data)
  },

  // 切换传承人状态
  toggleStatus: (id: number) => {
    return request.patch(`/api/v1/social/heritage/inheritors/${id}/status`)
  },

  // 删除传承人
  deleteInheritor: (id: number) => {
    return request.delete(`/api/v1/social/heritage/inheritors/${id}`)
  }
} 

File: backend/app/models/social.py
from typing import List, Optional
from enum import Enum
from datetime import datetime

from sqlalchemy import String, Text, Integer, Boolean, DateTime, ForeignKey, Enum as SQLEnum
from sqlalchemy.orm import Mapped, mapped_column, relationship

from .base import Base

class PostType(str, Enum):
    """动态类型枚举"""
    TEXT = "text"           # 文字动态
    IMAGE = "image"         # 图片动态
    VIDEO = "video"         # 视频动态
    DANCE = "dance"         # 舞蹈分享
    HERITAGE = "heritage"   # 非遗传承

class Post(Base):
    """动态模型"""
    __tablename__ = "posts"

    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    title: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    post_type: Mapped[PostType] = mapped_column(SQLEnum(PostType), default=PostType.TEXT)
    media_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    thumbnail_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    likes_count: Mapped[int] = mapped_column(Integer, default=0)
    comments_count: Mapped[int] = mapped_column(Integer, default=0)
    shares_count: Mapped[int] = mapped_column(Integer, default=0)
    is_public: Mapped[bool] = mapped_column(Boolean, default=True)
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False)  # 是否精选
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # 关系定义
    user: Mapped["User"] = relationship("User", back_populates="posts")
    comments: Mapped[List["PostComment"]] = relationship("PostComment", back_populates="post", cascade="all, delete-orphan")
    likes: Mapped[List["PostLike"]] = relationship("PostLike", back_populates="post", cascade="all, delete-orphan")

class PostComment(Base):
    """动态评论模型"""
    __tablename__ = "post_comments"

    post_id: Mapped[int] = mapped_column(Integer, ForeignKey("posts.id"), nullable=False)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    parent_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("post_comments.id"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    # 关系定义
    post: Mapped["Post"] = relationship("Post", back_populates="comments")
    user: Mapped["User"] = relationship("User", back_populates="post_comments")
    parent: Mapped[Optional["PostComment"]] = relationship("PostComment", remote_side="PostComment.id", back_populates="replies")
    replies: Mapped[List["PostComment"]] = relationship("PostComment", back_populates="parent")

class PostLike(Base):
    """动态点赞模型"""
    __tablename__ = "post_likes"

    post_id: Mapped[int] = mapped_column(Integer, ForeignKey("posts.id"), nullable=False)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    # 关系定义
    post: Mapped["Post"] = relationship("Post", back_populates="likes")
    user: Mapped["User"] = relationship("User", back_populates="post_likes")

class HeritageProject(Base):
    """非遗传承项目模型"""
    __tablename__ = "heritage_projects"

    name: Mapped[str] = mapped_column(String(200), nullable=False)
    description: Mapped[str] = mapped_column(Text, nullable=False)
    origin_location: Mapped[str] = mapped_column(String(100), nullable=False)  # 发源地
    category: Mapped[str] = mapped_column(String(50), nullable=False)  # 类别
    level: Mapped[str] = mapped_column(String(50), nullable=False)  # 保护级别
    cover_image: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    video_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    history: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # 历史背景
    characteristics: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # 特色特点
    inheritor_id: Mapped[Optional[int]] = mapped_column(Integer, ForeignKey("heritage_inheritors.id"), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # 关系定义
    inheritor: Mapped[Optional["HeritageInheritor"]] = relationship("HeritageInheritor", back_populates="projects")

class HeritageInheritor(Base):
    """非遗传承人模型"""
    __tablename__ = "heritage_inheritors"

    name: Mapped[str] = mapped_column(String(100), nullable=False)
    gender: Mapped[str] = mapped_column(String(10), nullable=False)
    birth_year: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    hometown: Mapped[str] = mapped_column(String(100), nullable=False)
    specialty: Mapped[str] = mapped_column(String(200), nullable=False)  # 专长
    biography: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # 个人简介
    achievements: Mapped[Optional[str]] = mapped_column(Text, nullable=True)  # 主要成就
    avatar: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    contact_info: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # 关系定义
    projects: Mapped[List["HeritageProject"]] = relationship("HeritageProject", back_populates="inheritor") 